File: mvp/src/main/java/com/mandria/android/mvp/PresenterCache.java
package com.mandria.android.mvp;

import java.util.HashMap;

/**
 * <p>A cache for the presenter.</p>
 * <p>This class should be a singleton or used with a dependency injector to be injected as an application singleton.</p>
 */
public final class PresenterCache {

    private static final String TAG = "PresenterCache";

    private HashMap<String, Presenter> mIdToPresenter;

    private HashMap<Presenter, String> mPresenterToId;

    /**
     * Constructor.
     */
    public PresenterCache() {
        mIdToPresenter = new HashMap<>();
        mPresenterToId = new HashMap<>();
    }

    /**
     * Gets a presenter using its id.
     *
     * @param id  Presenter id.
     * @param <P> Presenter type.
     * @return The presenter of the corresponding id.
     */
    @SuppressWarnings("unchecked")
    <P> P getPresenter(String id) {
        return (P) mIdToPresenter.get(id);
    }

    /**
     * Saves a presenter.
     *
     * @param presenter Presenter to save.
     */
    void savePresenter(Presenter presenter) {
        String id = presenter.getClass().getSimpleName() + "/" + System.nanoTime() + "/" + (int) (Math.random() * Integer.MAX_VALUE);

        MVPLogger.d(TAG, String.format("Saving presenter %s to cache with id %s", presenter.getClass().getSimpleName(), id));

        mIdToPresenter.put(id, presenter);
        mPresenterToId.put(presenter, id);
    }

    /**
     * Gets the id of a presenter, or null if presenter is not cached.
     *
     * @param presenter Presenter to retrieve the id from.
     * @return The presenter id or null.
     */
    String getId(Presenter presenter) {
        return mPresenterToId.get(presenter);
    }

    /**
     * Removes a presenter from the cache.
     *
     * @param presenter Presenter to remove.
     */
    void removePresenter(Presenter presenter) {
        MVPLogger.d(TAG, String.format("Removing presenter %s from cache", presenter.getClass().getSimpleName()));

        mIdToPresenter.remove(mPresenterToId.get(presenter));
    }

    /**
     * Removes all the presenter from the cache.
     */
    public void clear() {
        mPresenterToId.clear();
        mIdToPresenter.clear();
    }
}

File: core_kotlin/src/main/kotlin/com/mandria/android/mvp/PresenterCache.kt
package com.mandria.android.mvp


/**
 * A cache for the presenters.
 * This class should be a singleton or used with a dependency injector to be injected as an application singleton.
 */
class PresenterCache {

    private val idToPresenter: HashMap<String, Presenter<*>> = HashMap()

    private val presenterToId: HashMap<Presenter<*>, String> = HashMap()

    operator fun get(id: String) = idToPresenter[id]

    /**
     * Gets a presenter using its id.
     *
     * @param id  Presenter id.
     * @param <P> Presenter type.
     * @return The presenter of the corresponding id.
     */
    @Suppress("UNCHECKED_CAST")
    fun <P> getPresenter(id: String): P = idToPresenter[id] as P

    /** Saves a [presenter] in the cache. **/
    fun savePresenter(presenter: Presenter<*>) {
        val id =
            "${presenter.javaClass.simpleName}/${System.nanoTime()}/${(Math.random() * Int.MAX_VALUE).toInt()}"

        MVPLogger.d(TAG, "Saving presenter ${presenter.javaClass.simpleName} to cache with id $id")

        idToPresenter[id] = presenter
        presenterToId[presenter] = id
    }

    /**
     * Gets the id of a [presenter], or null if [presenter] is not cached.
     *
     * @return The presenter id or null.
     */
    fun getId(presenter: Presenter<*>): String? {
        return presenterToId[presenter]
    }

    /** Removes a [presenter] from the cache. **/
    fun removePresenter(presenter: Presenter<*>) {
        MVPLogger.d(TAG, "Removing presenter ${presenter.javaClass.simpleName} from cache")
        idToPresenter.remove(presenterToId.remove(presenter))
    }

    /** Removes all the presenter from the cache. **/
    fun clear() {
        presenterToId.clear()
        idToPresenter.clear()
    }

    companion object {
        const val TAG = "PresenterCache"
    }
}